pkg/chunk/local/fs_object_client.go

```diff
@@ -53,57 +52,16 @@ func NewFSObjectClient(cfg FSConfig) (*FSObjectClient, error) {
 // Stop implements ObjectClient
 func (FSObjectClient) Stop() {}
 
-// PutChunks implements ObjectClient
-func (f *FSObjectClient) PutChunks(ctx context.Context, chunks []chunk.Chunk) error {
-	for i := range chunks {
-		buf, err := chunks[i].Encoded()
-		if err != nil {
-			return err
-		}
-
-		filename := base64.StdEncoding.EncodeToString([]byte(chunks[i].ExternalKey()))
-		if err := f.PutObject(ctx, filename, bytes.NewReader(buf)); err != nil {
-			return err
-		}
-	}
-	return nil
-}
-
-// GetChunks implements ObjectClient
-func (f *FSObjectClient) GetChunks(ctx context.Context, chunks []chunk.Chunk) ([]chunk.Chunk, error) {
-	return util.GetParallelChunks(ctx, chunks, f.getChunk)
-}
-
-func (f *FSObjectClient) getChunk(ctx context.Context, decodeContext *chunk.DecodeContext, c chunk.Chunk) (chunk.Chunk, error) {
-	filename := base64.StdEncoding.EncodeToString([]byte(c.ExternalKey()))
-
-	readCloser, err := f.GetObject(ctx, filename)
-	if err != nil {
-		return chunk.Chunk{}, err
-	}
-
-	defer readCloser.Close()
-
-	buf, err := ioutil.ReadAll(readCloser)
-	if err != nil {
-		return chunk.Chunk{}, err
-	}
-
-	if err := c.Decode(decodeContext, buf); err != nil {
-		return c, err
-	}
-
-	return c, nil
-}
-
-// Get object from the store
+// GetObject from the store
 func (f *FSObjectClient) GetObject(ctx context.Context, objectKey string) (io.ReadCloser, error) {
-	return os.Open(path.Join(f.cfg.Directory, objectKey))
+	filename := base64.StdEncoding.EncodeToString([]byte(objectKey))
```

Comment:
I'm not sure doing the base64 encoding here (in the `GetObject()`) is correct. When Sandeep submitted the former PR to introduce `GetObject()` the idea was to intentionally keep it generic so that `objectKey` can contain `/` to navigate through sub directories. We should keep this behavior (and also add a unit test to protect from this regression).

Reply:
I personally don't like using the base64 encoding either. However, the current implementation uses it to encode chunks. Maybe I can maintain this behavior with a config option to allow for local storage users to keep base64 encoded chunk names. I personally think storing chunks locally using the chunk external key would work great, but I haven't thought through all of the implications. Having each users chunks stored in a directory seems conceptually sound to me:

```go
// ExternalKey returns the key you can use to fetch this chunk from external
// storage. For newer chunks, this key includes a checksum.
func (c *Chunk) ExternalKey() string {
	// Some chunks have a checksum stored in dynamodb, some do not.  We must
	// generate keys appropriately.
	if c.ChecksumSet {
		// This is the inverse of parseNewExternalKey.
		return fmt.Sprintf("%s/%x:%x:%x:%x", c.UserID, uint64(c.Fingerprint), int64(c.From), int64(c.Through), c.Checksum)
	}
	// This is the inverse of parseLegacyExternalKey, with "<user id>/" prepended.
	// Legacy chunks had the user ID prefix on s3/memcache, but not in DynamoDB.
	// See comment on parseExternalKey.
	return fmt.Sprintf("%s/%d:%d:%d", c.UserID, uint64(c.Fingerprint), int64(c.From), int64(c.Through))
}
```

Reply:
> I personally think storing chunks locally using the chunk external key would work great, but I haven't thought through all of the implications.

If we remove the base64 we would end up with a breaking change because `base64(chunk.ExternalKey()) != chunk.ExternalKey()`.

I think this PR should be a mere refactoring PR, without altering the behaviour in any way, so `PutChunks()` and `GetChunks()` should base64 encode the key (when using the FS client), while `PutObject()` and `GetObject()` not.

An option to achieve it may be adding an extra function to the `chunk.Client` interface like `EncodeExternalKey(key string) string`. This function is a noop for all clients, except the FS client which will do the base 64. Finally, `PutChunks()` and `GetChunks` will call `EncodeExternalKey(chunks[i].ExternalKey())`.

Reply:
I think an easy way to handle this is to inject an optional `Encoder` function into the object store chunk client.

```go
// KeyEncoder is used to encode chunk keys before writing/retrieving chunks
// from the underlying ObjectClient
type KeyEncoder func(string) string

// Client is used to store chunks in object store backends
type Client struct {
	store      ObjectClient
	keyEncoder KeyEncoder
}
```

In `PutChunks`:
```go
		key := chunks[i].ExternalKey()
		if o.keyEncoder != nil {
			key = o.keyEncoder(key)
		}
```

in `getChunk`:
```go
func (o *Client) getChunk(ctx context.Context, decodeContext *DecodeContext, c Chunk) (Chunk, error) {
	key := c.ExternalKey
	if o.keyEncoder != nil {
		key = o.keyEncoder(key)
	}
```

Reply:
Yes, even better.